### DISCUSSION TITLE
blog/react-19-and-suspense-a-drama-in-3-acts

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#106] [Announcements]
# blog/react-19-and-suspense-a-drama-in-3-acts

React 19 is a very promising release - but there's something not quite right yet with suspense...

https://tkdodo.eu/blog/react-19-and-suspense-a-drama-in-3-acts

**Comment by devinrhode2** [comment]
For users that have not yet upgraded to React 18… should we just hold off for a few more months? What percent of the v18 upgrade work would be wasted?

**Reply by TkDodo** [reply]
things work fine in React 18.

**Reply by SeanCassiere** [reply]
@devinrhode2 You'll be in a better spot if you upgrade to react@18.3.1 as per the React core team's suggestion, since this this release more closely mirrors what's expected to come in React 19 with the necessary warnings in-place for APIs that are going away.
Then once 19 comes out, it'll be a matter of yanking out `<Context.Provider>` for `<Context>` and switching from `React.forwardRef` to `ref` as a prop... and ofc, you can choose to use [everything else](https://react.dev/blog/2024/04/25/react-19-upgrade-guide) with the release as well.

**Reply by devinrhode2** [reply]
@SeanCassiere so it sounds like all of the v18 upgrade work will be necessary or useful in getting to the v19 promised land?

(iirc the ref and context APIs didn't change from 16.8->18)

**Reply by devinrhode2** [reply]
So v19 upgrade is basically:
- Everything in v18 upgrade
- ref and context APIs
?

**Reply by SeanCassiere** [reply]
> @SeanCassiere so it sounds like all of the v18 upgrade work will be necessary or useful in getting to the v19 promised land?
> 
> (iirc the ref and context APIs didn't change from 16.8->18)

Yup! Imo the React 18 upgrade is quite useful.

Plus, pretty much all your deps in your package.json will have versions out that are *fully* compatible with React 18. When it comes time to make the jump to React 19 and something gets borked, it'll be much easier finding out if it is a package that is not yet compatible with 19 or if it is something wonky in how you wrote React in your App code.

**Comment by vitormarkis** [comment]
In the `expensive-sibling` example. It only makes sense because `ExpensiveComponent` comes right after a component that suspended. If the suspended component comes after the expensive one, the render would still be thrown away once it reaches the SuspendingComponent, right?

I don't get it, they're assuming you'll always have the suspending components early in the JSX order.

**Reply by TkDodo** [reply]
If your suspending component is last then yes, it likely doesn't make a difference. But I think there are enough cases where there are siblings that got unnecessarily pre-rendered in v18. It seems to have been a big enough issue at facebook to warrant this change.

**Comment by s5unanow** [comment]
Thanks for breaking this down with great examples. Now I much more aware of the situation!

**Comment by everdimension** [comment]
Really amazing work on your side, I think it's underestimated how much what you did impacts the community.

This part is interesting though
> Hoisting your data requirements is a good idea regardless of how suspense works, and I also recommend doing that

It's not always better, though, is it? I think it's a good solution with _a lot of_ trade-offs. And I understand that it's probably the best we can do at the moment.
But... it's so far from an ideal one. Hoisting data requirements means that your code is split between the data fetching and the component that needs that data. You lose colocation, which, in turn, can lead to overfetching and errors. The former can happen when you no longer render the component you fetched data for. The latter can happen when the API changes. I guess the errors are easier to prevent with clever typings, but potential overfetching will still be an issue, will it not?

**Reply by TkDodo** [reply]
> You lose colocation, which, in turn, can lead to overfetching and errors.

depends on how you do it really. With relay, you'll get all of this automatically because of their compiler. That's ideal I'd say.

With react-query, you won't get errors because you'll use the same `queryKey` and `queryFn` in both places, so that's great. Losing colocation is a trade-off and the overfetching part is real, but it's still better than not doing it imo, hence the recommendation.

**Comment by prateek3255** [comment]
First of all thanks for summarizing all of this it makes a lot of sense now. I just had one confusion about the definition of `fetch-on-render` vs `render-as-you-fetch`. I always thought `fetch-on-render` meant where you fetch on mount, like in `useEffect` and with `render-as-you-fetch` you throw the promise while rendering that initiates data fetching as well. Something like this:

```jsx
import React, { Suspense, useState } from 'react';

// A function to fetch data with a promise
const fetchData = () => {
  return fetch('/api/data').then(response => response.json());
};

// A custom hook that uses Suspense to fetch data
const useData = () => {
  const [data, setData] = useState(null);

  if (!data) {
    throw fetchData().then(setData);
  }

  return data;
};

// A component that uses the custom hook
function MyComponent() {
  const data = useData();
  return <div>Data: {JSON.stringify(data)}</div>;
}

// Main app component with Suspense
function App() {
  return (
    <Suspense fallback={<div>Loading...</div>}>
      <MyComponent />
    </Suspense>
  );
}

export default App;
```

Now from your blog what I understand is in the render as you fetch model you initiate the fetching earlier even before the component starts rendering and and then consume that promise in the suspended component. Was my understanding wrong earlier?

**Reply by TkDodo** [reply]
the names are confusing, but as far as I know, `render-as-you-fetch` implies that you start the fetch before rendering. your examples are just `fetch-on-render` in two different flavours: effects (pretty much what `useQuery` does) and suspense (pretty much what `useSuspenseQuery` does).

**Comment by zeorin** [comment]
I feel like this is solvable in a way that we can all have our cake and eat it, too:

`<SuspendingComponent />` can trigger suspense, and this can trigger the fallback. I don't see a reason why it can't trigger the fallback before `<ExpensiveComponent />` has rendered, I thought Fibre is supposed to enable this kind of re-prioritization of chunks of work... And, I also don't see why its result needs to be thrown away… They could keep it and just reuse it if on resume its inputs haven't changed, possibly they can decide to do that only if it's memoed, i.e. just re-render the thing but keep its old result around in case it's memoized and should be reused.

**Reply by TkDodo** [reply]
rendering of children happens top-down, because children is an array. if `<SuspendingComponent />` comes first and triggers suspense, that means it literally `throws` (which stops execution), and the nearest suspense boundary picks it up. That means it never gets to render `<ExpensiveComponent />` if it shows the fallback immediately. What could be done is schedule a special, low-prio render pass "in the background", and that is something the react team considers. But it's not implemented.

**Reply by zeorin** [reply]
Yes, I understand, I've written some raw suspense triggering code myself (by throwing a promise to suspend and using `useSyncExternalStore` to trigger a new render when ready).

I also understand why it doesn't (or rather, didn't) work in React 19, as they changed the execution model, but my point is that AFAICT there's no inherent theoretical limitation that precludes the approach I outlined above, since Fiber introduced chunking React's work, using cooperative multitasking, which also allows for re-prioritisation of Fibers, as has been suggested when talking about e.g. progressive hydration.

Perhaps they could achieve this with some of the work that's being done for Activities (formerly Offscreen Canvas)?

**Comment by ebceu4** [comment]
Could &lt;Suspense&gt; and &lt;SuspenseWaterfall&gt; coexist to allow developers to choose the strategy for their particular case?

**Reply by TkDodo** [reply]
seems like a meh api

**Reply by JonahPlusPlus** [reply]
That was my first thought. Why change the behavior of an existing component instead of introducing a new one?

Edit: Or just expose a property for controlling the behavior. So like:
```jsx
<Suspense order="serial"> // default: order="parallel"
...
</Suspense>
```

**Comment by whyte25** [comment]
This is impressive, welldone @TkDodo

**Comment by IVLIU** [comment]
I'm wondering if it is possible to trigger an urgent update to render the fallback immediately after collecting the first promise, and then trigger a transition update to continue collecting promises immediately afterward.

**Comment by deanrad** [comment]
React was born of a hypothesis that declarative programming was enough for UIs - but here you are struggling to get your intentions implemented. The fault is in the paradigm - if users had explicit control over when things happened, they would not find their apps breaking when the 'magic' is rewritten by the framework arbitrarily..

For my taste - do all async in pure RxJS/RxFx - then just hand state trees down into memoized, hookless, functional components - and don't worry about what terrible semantics lie inside React's async choices - just control yours in a language/paradigm built for that, not just for DOM trees.

**Comment by Akshay090** [comment]
whats the status of this at present after a year, is it waterfall or parallel 😅

**Reply by TkDodo** [reply]
It's parallel 👍

**Reply by Akshay090** [reply]
daimm, crazy fast reply. 

I was reading up on this again to understand the current state of Suspense.  
Here’s a quick TL;DR of the lore for anyone stumbling on this later like me:

### 🟦 React 18 (old behaviour)
- When the first sibling suspends, React **continues rendering** the next sibling.
- Both siblings start their data fetches **in parallel**.
- React commits the fallback **only after** attempting to render the second sibling — so the fallback appears a bit later.
- **Result:** Parallel fetches.

### 🟧 React 19 Beta (around May 2024)
- As soon as the first sibling suspends, React **commits the fallback immediately**.
- It **does not render** the next sibling until the first one resolves.
- This causes the second sibling’s fetch to start **after** the first one — a **waterfall**.
- This change broke expectations and received community pushback.
- **Result:** Waterfall / serial fetches.

### 🟩 React 19 Final (current behaviour)
- On first suspension, React **commits the fallback immediately** (good UX).
- **But React also continues rendering the next sibling**, so its data fetch can start right away.
- This restores **parallel** fetch behaviour, similar to React 18, but with React 19’s UX improvements.
- **Result:** Parallel fetches + faster fallback commit.

### 🔗 CodeSandbox PoCs
- **React 19 Final:**  
  https://codesandbox.io/p/sandbox/react-19-suspense-behaviour-63nwnc
- **React 18:**  
  https://codesandbox.io/p/sandbox/n7m623
- **React 19 Beta:**  
  https://codesandbox.io/p/devbox/react-19-suspense-behaviour-forked-cy33lw?workspaceId=ws_PMifpD44PW8jzHHYtjRpfY

**Reply by s5unanow** [reply]
Great findings! Much appreciate it.

**Comment by kein-1** [comment]
I wonder if i'm doing something wrong or what but Suspense on a global level does not apply. 


i.e The following Fallback component won't run;

```
export const Route = createRootRouteWithContext<RouterContext>()({
  component: () => (
    <Suspense fallback={<Fallback />}>
      <div className="flex min-h-svh w-full flex-col">
        <Outlet />
        <TanStackRouterDevtools position="bottom-right" />
        <Footer />
      </div>
    </Suspense>
  ),
  notFoundComponent: NotFound,
  errorComponent: ErrorHandling,
})
```

BUT this will:

```
export const Route = createFileRoute({
  component: () => (
    <Suspense fallback={<Fallback />}> // i see fallback, but if i took out the <Suspense/> wrapping, i don't see anything
      <RouteComponent />
    </Suspense>
  ),
  loader: async ({ context }) => {
    console.log('running loader...')
    context.queryClient.ensureQueryData({
      queryKey: ['test-suspense'],
      queryFn: fetchSomething,
    })
    console.log('loader done')
  },
})

```


For suspense and loader: do we need to wrap it at the router component level..?

**Reply by TkDodo** [reply]
TanStack Router wraps each route per default in a suspense boundary, and it will render what you provide as `pendingComponent` as a fallback. So the simplest solution is likely:

```
notFoundComponent: NotFound,
errorComponent: ErrorHandling,
pendingComponent: Fallback,
```
